File: atc-python/atc_sdk/keys.py
"""ATC/1.0 key utilities — Ed25519 keypair generation and signing."""

from __future__ import annotations

from cryptography.hazmat.primitives.asymmetric.ed25519 import Ed25519PrivateKey, Ed25519PublicKey
from cryptography.hazmat.primitives import serialization
import base64
import json
from dataclasses import dataclass
from typing import Tuple

ATC_ALGORITHM = "Ed25519"


@dataclass
class KeyPair:
    """A CA or agent keypair with Ed25519 keys."""
    public_key: str  # base64 SPKI
    private_key: str  # base64 PKCS8
    raw_private_key: Ed25519PrivateKey
    raw_public_key: Ed25519PublicKey


def generate_keypair() -> KeyPair:
    """Generate a new Ed25519 keypair."""
    private_key = Ed25519PrivateKey.generate()
    public_key = private_key.public_key()

    priv_pem = private_key.private_bytes(
        encoding=serialization.Encoding.DER,
        format=serialization.PrivateFormat.PKCS8,
        encryption_algorithm=serialization.NoEncryption(),
    )
    pub_pem = public_key.public_bytes(
        encoding=serialization.Encoding.DER,
        format=serialization.PublicFormat.SubjectPublicKeyInfo,
    )

    return KeyPair(
        public_key=base64.b64encode(pub_pem).decode("ascii"),
        private_key=base64.b64encode(priv_pem).decode("ascii"),
        raw_private_key=private_key,
        raw_public_key=public_key,
    )


def load_keypair_from_private(base64_private_key: str) -> KeyPair:
    """Reconstruct a keypair from a saved base64 PKCS8 private key."""
    priv_der = base64.b64decode(base64_private_key)
    private_key = serialization.load_der_private_key(priv_der, password=None)
    public_key = private_key.public_key()

    pub_pem = public_key.public_bytes(
        encoding=serialization.Encoding.DER,
        format=serialization.PublicFormat.SubjectPublicKeyInfo,
    )

    return KeyPair(
        public_key=base64.b64encode(pub_pem).decode("ascii"),
        private_key=base64_private_key,
        raw_private_key=private_key,
        raw_public_key=public_key,
    )


def sign_message(message: bytes | str, private_key: Ed25519PrivateKey) -> str:
    """Sign a message with Ed25519, return base64 signature."""
    if isinstance(message, str):
        message = message.encode("utf-8")
    sig = private_key.sign(message)
    return base64.b64encode(sig).decode("ascii")


def verify_signature(message: bytes | str, base64_signature: str, base64_public_key: str) -> bool:
    """Verify an Ed25519 signature."""
    try:
        pub_der = base64.b64decode(base64_public_key)
        public_key = serialization.load_der_public_key(pub_der)
        if isinstance(message, str):
            message = message.encode("utf-8")
        sig = base64.b64decode(base64_signature)
        public_key.verify(sig, message)  # raises on invalid
        return True
    except Exception:
        return False

File: atc-rust/src/keys.rs
//! ATC/1.0 key utilities — Ed25519 keypair generation

use base64::Engine;
use ed25519_dalek::{SigningKey, VerifyingKey};
use ed25519_dalek::pkcs8::{DecodePrivateKey, EncodePrivateKey, EncodePublicKey, spki::Error as SpkiError};
use rand::rngs::OsRng;
use serde::{Serialize, Deserialize};
use thiserror::Error;

#[derive(Debug, Error)]
pub enum KeyError {
    #[error("base64 decode error: {0}")]
    Base64(#[from] base64::DecodeError),
    #[error("SPKI/PKCS8 error: {0}")]
    Spki(#[from] SpkiError),
    #[error("signature error: {0}")]
    Signature(#[from] ed25519_dalek::SignatureError),
}

/// A CA or agent keypair with Ed25519 keys.
#[derive(Serialize, Deserialize, Clone)]
pub struct KeyPair {
    /// Base64-encoded full SPKI (44 DER bytes → 60 base64 chars)
    pub public_key: String,
    /// Base64-encoded full PKCS8 (48 DER bytes → 64 base64 chars)
    pub private_key: String,
    /// The signing key (for signing operations)
    #[serde(skip)]
    pub raw_signing_key: SigningKey,
    /// The verifying key (for verifying operations)
    #[serde(skip)]
    pub raw_verifying_key: VerifyingKey,
}

/// Generate a new Ed25519 keypair.
pub fn generate_keypair() -> Result<KeyPair, KeyError> {
    let mut csprng = OsRng;
    let signing_key = SigningKey::generate(&mut csprng);
    let verifying_key = signing_key.verifying_key();

    let priv_pem = signing_key.to_pkcs8_der()?;
    let pub_pem = verifying_key.to_public_key_der()?;

    Ok(KeyPair {
        public_key: base64::engine::general_purpose::STANDARD.encode(pub_pem.as_bytes()),
        private_key: base64::engine::general_purpose::STANDARD.encode(priv_pem.as_bytes()),
        raw_signing_key: signing_key,
        raw_verifying_key: verifying_key,
    })
}

/// Reconstruct a keypair from a saved base64 PKCS8 private key.
pub fn load_keypair_from_private(b64_private_key: &str) -> Result<KeyPair, KeyError> {
    let priv_der = base64::engine::general_purpose::STANDARD.decode(b64_private_key)?;
    let signing_key = SigningKey::from_pkcs8_der(&priv_der)?;
    let verifying_key = signing_key.verifying_key();

    let pub_pem = verifying_key.to_public_key_der()?;

    Ok(KeyPair {
        public_key: base64::engine::general_purpose::STANDARD.encode(pub_pem.as_bytes()),
        private_key: b64_private_key.to_string(),
        raw_signing_key: signing_key,
        raw_verifying_key: verifying_key,
    })
}

/// Sign a message with an Ed25519 signing key. Returns base64 signature.
pub fn sign_message(message: &[u8], signing_key: &SigningKey) -> String {
    let sig = signing_key.sign(message);
    base64::engine::general_purpose::STANDARD.encode(sig.to_bytes())
}

/// Verify an Ed25519 signature.
pub fn verify_signature(message: &[u8], b64_signature: &str, b64_public_key: &str) -> bool {
    use ed25519_dalek::Verifier;
    let sig_bytes = match base64::engine::general_purpose::STANDARD.decode(b64_signature) {
        Ok(b) => b,
        Err(_) => return false,
    };
    let sig_array: [u8; 64] = match sig_bytes.as_slice().try_into() {
        Ok(a) => a,
        Err(_) => return false,
    };
    let sig = ed25519_dalek::Signature::from_bytes(&sig_array);

    let pub_der = match base64::engine::general_purpose::STANDARD.decode(b64_public_key) {
        Ok(b) => b,
        Err(_) => return false,
    };
    let verifying_key = match VerifyingKey::from_public_key_der(&pub_der) {
        Ok(k) => k,
        Err(_) => return false,
    };

    verifying_key.verify(message, &sig).is_ok()
}
